controllers/awsmachine_controller_unit_test.go

```diff
@@ -17,25 +17,1507 @@ limitations under the License.
+				g := NewWithT(t)
+				awsMachine := getAWSMachine()
+				setup(t, g, awsMachine)
+				defer teardown(t, g)
+				runningInstance(t, g)
+				_, _ = reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+
+				g.Expect(len(ms.AWSMachine.Finalizers)).To(Equal(0))
+			})
+		})
+
+		t.Run("when there's a provider ID", func(t *testing.T) {
+			id := providerID
+			providerID := func(t *testing.T, g *WithT) {
+				t.Helper()
+				_, err := noderefutil.NewProviderID(id)
+				g.Expect(err).To(BeNil())
+
+				ms.AWSMachine.Spec.ProviderID = &id
+			}
+
+			t.Run("should look up by provider ID when one exists", func(t *testing.T) {
+				g := NewWithT(t)
+				awsMachine := getAWSMachine()
+				setup(t, g, awsMachine)
+				defer teardown(t, g)
+
+				providerID(t, g)
+				expectedErr := errors.New("no connection available ")
+				ec2Svc.EXPECT().InstanceIfExists(PointsTo("myMachine")).Return(nil, expectedErr)
+
+				_, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+				g.Expect(errors.Cause(err)).To(MatchError(expectedErr))
+			})
+
+			t.Run("should try to create a new machine if none exists and add finalizers", func(t *testing.T) {
+				g := NewWithT(t)
+				awsMachine := getAWSMachine()
+				setup(t, g, awsMachine)
+				defer teardown(t, g)
+
+				providerID(t, g)
+				expectedErr := errors.New("Invalid instance")
+				ec2Svc.EXPECT().InstanceIfExists(gomock.Any()).Return(nil, nil)
+				secretSvc.EXPECT().Create(gomock.Any(), gomock.Any()).Return("test", int32(1), nil).Times(1)
+				ec2Svc.EXPECT().CreateInstance(gomock.Any(), gomock.Any()).Return(nil, expectedErr)
+				secretSvc.EXPECT().UserData(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil).Times(1)
+
+				_, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+				g.Expect(ms.AWSMachine.Finalizers).To(ContainElement(infrav1.MachineFinalizer))
+				g.Expect(errors.Cause(err)).To(MatchError(expectedErr))
+			})
+		})
+
+		t.Run("should fail to find instance if no provider ID provided", func(t *testing.T) {
+			g := NewWithT(t)
+			awsMachine := getAWSMachine()
+			setup(t, g, awsMachine)
+			defer teardown(t, g)
+			id := "aws////myMachine"
+
+			ms.AWSMachine.Spec.ProviderID = &id
+			expectedErr := "providerID must be of the form <cloudProvider>://<optional>/<segments>/<provider id>"
+
+			_, err := reconciler.reconcileNormal(context.Background(), ms, cs, cs, cs)
+			g.Expect(err.Error()).To(ContainSubstring(expectedErr))
+		})
+
+		t.Run("when instance creation succeeds", func(t *testing.T) {
+			var instance *infrav1.Instance
+
+			instanceCreate := func(t *testing.T, g *WithT) {
+				t.Helper()
+
+				instance = &infrav1.Instance{
+					ID:        "myMachine",
+					VolumeIDs: []string{"volume-1", "volume-2"},
+				}
+				instance.State = infrav1.InstanceStatePending
+
+				ec2Svc.EXPECT().GetRunningInstanceByTags(gomock.Any()).Return(nil, nil)
```

Comment:
Shouldn't this return the instance itself if the creation is successful? This is the same in few other places.

Reply:
It is returning instance, it's just that we have used a global variable `instance`(local to `when instance creation succeeds` block) which is referenced in each sub test case